packages/core/lib/settings/asset_type.dart

```diff
@@ -48,30 +50,157 @@ class AssetType {
   void addChild(AssetType type) {
     _children.add(type);
   }
+
+  /// Returns a name for this asset.
+  String get name {
+    return withoutExtension(path);
+  }
 }
 
-class AssetTypeIsUniqueWithoutExtension {
-  AssetTypeIsUniqueWithoutExtension({
-    required this.assetType,
-    required this.isUniqueWithoutExtension,
-  });
+/// Represents a AssetType with modifiers on it to mutate the [name] to ensure
+/// it is unique in a larger list and a valid dart identifer.
+///
+/// See [AssetTypeIterable.mapToUniqueAssetType] for the algorithm.
+class UniqueAssetType extends AssetType {
+  UniqueAssetType({
```

Comment:
Anything prevent this class from being immutable?

Reply:
This class international gets mutated as collisions are found. I'm all for immutable data structures, but changing this would complicate the flow elsewhere (specifically having to add remove from lists while iterating)